infix/parser.go

```diff
@@ -0,0 +1,224 @@
+package infix
+
+import (
+	"fmt"
+	"strconv"
+)
+
+type parser struct {
+	tokens chan Tokval
+	pocket *Tokval // lookahead
+}
+
+var TokEOF = Tokval{
+	T: EOF,
+	V: "<eof>",
+}
+
+func parserErr(tok Tokval) error {
+	return fmt.Errorf("unexpected %s", tok)
+}
+
+func Parse(code string) (Node, error) {
+	p := &parser{
+		tokens: Lex(code),
+	}
+	return p.parseExpr()
+}
+
+func (p *parser) peek() Tokval {
+	if p.pocket != nil {
+		val := *p.pocket
+		p.pocket = nil
+		return val
+	}
+
+	val, ok := <-p.tokens
+	if !ok {
+		return TokEOF
+	}
+	p.pocket = &val
+	return val
+}
+
+func (p *parser) next() Tokval {
+	if p.pocket != nil {
+		val := *p.pocket
+		p.pocket = nil
+		return val
+	}
+	tok, ok := <-p.tokens
+	if !ok {
+		return TokEOF
+	}
+	return tok
+}
+
+func (p *parser) parseExpr() (Node, error) {
+	var eoferr = func(expect string) error {
+		return fmt.Errorf("premature eof, expects %s",
+			expect)
+	}
+
+	tok := p.peek()
+	if tok.T == EOF {
+		return nil, eoferr("expression")
+	}
+
+	// left hand side of expr
```

Comment:
Instead of two comments "left hand side" and "right hand side" we could have two functions

Reply:
yes, I commonly implement this way first to visualize *what* makes sense to extract in function and *how*. My brain is very limited.
Will improve.